File: Lab1_3/2dihot.py
# Метод дихотомии (половинного деления)
import math

# Функция
def func(x):
    return math.exp(-x) - math.sqrt(x - 1)

# Метод половинного деления для нахождения корня уравнения
def find_root(a, b, eps):
    while abs(b - a) > eps:
        mid = (a + b) / 2
        if func(mid) == 0:
            return mid
        elif func(a) * func(mid) < 0:
            b = mid
        else:
            a = mid

    return (a + b) / 2

# Начальные границы интервала и требуемая точность
a = 1.0
b = 3.0
epsilon = 1e-6


root = find_root(a, b, epsilon)
print(f"Численный корень уравнения: {root}")

File: Lab1_3/2dihot.cpp
//Метод дихотомии (деления пополам)
#define _USE_MATH_DEFINES
#include <iostream>
#include <iomanip>
#include <cmath>
using namespace std;

double func(double x)
{
    return exp(-x) - sqrt(x - 1);
}
double find(double x0, double x1, double eps)
{
  double left = x0, right = x1, x, fl, fr, f;
  int iter = 0;
  cout << "a = " << x0 << "; b = " << x1 << "; ";
  do {
    x = (left + right) / 2;
    f = func(x);
    if (f > 0) right = x;
    else left = x;
    iter++;
  } while (fabs(f) > eps && iter<20000);
  cout <<"\nNumber of iterations is " << iter  << ". " << endl;
  return x;
}
int main() 
{
  double x0,x1,x,eps;
  cout << "a = ";
  cin >>x0;
  cout << "b = ";
  cin >>x1;
  cout << "eps = ";
  cin >>eps;
  x=find(x0,x1,eps);
  cout <<"Root x= "<< fixed << setprecision(16) << x;
  cout <<"\nFunction value f(x)="<< func(x);
  cin.get(); 
  return 0;
}